logging/src/lib.rs

```diff
@@ -151,108 +161,144 @@ impl<T> Logger<T> {
+    pub fn log<S: Into<T>>(&self, event: S)
+    where
+        CB: PushInto<(Duration, T)>,
+    {
+        self.inner.log(event.into());
+    }
+
+    /// Logs multiple events. Equivalent to [`Logger::log_many`], with the exception that it
+    /// converts the events to `T` before logging.
+    pub fn log_many<I>(&self, events: I)
+    where
+        I: IntoIterator, I::Item: Into<T>,
+        CB: PushInto<(Duration, T)>,
+    {
+        self.inner.log_many(events.into_iter().map(Into::into));
+    }
 }
 
-impl<T, A: ?Sized + FnMut(&Duration, &mut Vec<(Duration, T)>)> LoggerInner<T, A> {
-
-    /// The upper limit for buffers to allocate, size in bytes. [Self::buffer_capacity] converts
-    /// this to size in elements.
-    const BUFFER_SIZE_BYTES: usize = 1 << 13;
-
-    /// The maximum buffer capacity in elements. Returns a number between [Self::BUFFER_SIZE_BYTES]
-    /// and 1, inclusively.
-    // TODO: This fn is not const because it cannot depend on non-Sized generic parameters
-    fn buffer_capacity() -> usize {
-        let size =  ::std::mem::size_of::<(Duration, T)>();
-        if size == 0 {
-            Self::BUFFER_SIZE_BYTES
-        } else if size <= Self::BUFFER_SIZE_BYTES {
-            Self::BUFFER_SIZE_BYTES / size
-        } else {
-            1
+impl<CB: ContainerBuilder, T> Clone for TypedLogger<CB, T> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            _marker: PhantomData,
         }
     }
+}
+
+impl<CB: ContainerBuilder, T> From<Logger<CB>> for TypedLogger<CB, T> {
+    fn from(inner: Logger<CB>) -> Self {
+        TypedLogger {
+            inner,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<CB: ContainerBuilder, T> std::ops::Deref for TypedLogger<CB, T> {
+    type Target = Logger<CB>;
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
 
-    pub fn log_many<I>(&mut self, events: I)
-        where I: IntoIterator, I::Item: Into<T>
+impl<CB: ContainerBuilder, A: ?Sized + FnMut(&Duration, &mut CB::Container)> LoggerInner<CB, A> {
+    fn log_many<I>(&mut self, events: I)
+        where I: IntoIterator, CB: PushInto<(Duration, I::Item)>,
     {
         let elapsed = self.time.elapsed() + self.offset;
         for event in events {
-            self.buffer.push((elapsed, event.into()));
-            if self.buffer.len() == self.buffer.capacity() {
-                // Would call `self.flush()`, but for `RefCell` panic.
-                (self.action)(&elapsed, &mut self.buffer);
-                // The buffer clear could plausibly be removed, changing the semantics but allowing users
-                // to do in-place updates without forcing them to take ownership.
-                self.buffer.clear();
-                let buffer_capacity = self.buffer.capacity();
-                if buffer_capacity < Self::buffer_capacity() {
-                    self.buffer.reserve((buffer_capacity+1).next_power_of_two());
-                }
```

Comment:
We had logic in place that tried to grow the buffers geometrically, but I'm not sure that was still working as originally intended. With the switch to container builders, that logic certainly doesn't exist here anymore.